### ISSUE TITLE
Add a QueryCache implementation that caches generated SQL

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#8320]
Now that we can coerce the result type of a plain SQL `ResultQuery` (see #4473), we can cache the SQL generated from arbitrary jOOQ statements and run plain SQL queries instead. An example how this should feel in user code:

```java
class UserQueries extends QueryCache {
  Result<Record2<String, String>> getActorById(Integer... ids) {
    return cached("getActorById", ctx -> ctx
        .select(ACTOR.FIRST_NAME, ACTOR.LAST_NAME)
        .from(ACTOR)
        .where(ACTOR.ID.in(ids)), ids).fetch();
  }
}
```

The `cached()` method would look like this:

```java
<Q extends Query, R extends Record> Q cached(
  Object key, 
  Function<? super DSLContext, ? extends Q> query,
  Object... bindValues
);
```

The ugly part here is that bind values have to be repeated twice:

- Once in the query to generate the right SQL
- Once passed to the cached function, which binds the bind values to the cached plain SQL `ResultQuery`

**Comment by JanecekPetr** [comment]
FYI, there is a niche demand for this. Our use-case is using jOOQ for querying into in-process, in-memory DBs. In that scenario, jOOQ is taking up a significant portion of the query runtime.

As a solution, we're caching the rendered SQL in a way similar to what you're suggesting here. I personally do not like the code, but it helps a good amount. I do not see this as a core jOOQ feature, maybe as an additional artifact, jooq-query-cache? Either way, not a big priority. This exists just to let you know that the problem exists in the wild.

I also hoped for https://github.com/jOOQ/jOOQ/projects/2 to move in 3.14 for query caching and sharing purposes, but that seems to be a long way away now.

**Comment by lukaseder** [comment]
Thank you very much for your feedback, @JanecekPetr.

I'm sure the problem exists in the wild in the scenario that you mention. Every unfair benchmark out there uses H2 to compare various SQL APIs and ORMs, and in trivial `INSERT` or `SELECT` statements, jOOQ's SQL generation is really a bottleneck.

I'm still hoping for this to become a core jOOQ feature via https://github.com/jOOQ/jOOQ/projects/2. Imagine a query that can be stored in a static variable, and which will generate a cachable SQL string per configuration (dialect/settings) without you having to do anything.

> I also hoped for https://github.com/jOOQ/jOOQ/projects/2 to move in 3.14 for query caching and sharing purposes, but that seems to be a long way away now.

It's a huge project, and I don't think the majority of users will profit from it. The roadmap is driven in parts by the popularity of feature requests:
https://github.com/jOOQ/jOOQ/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

For jOOQ 3.14, Kotlin, embeddable types, SQL/XML and SQL/JSON were equally big projects that produce much more value. This doesn't mean that a new query object model wouldn't be great. It definitely would be, especially for folks who'd like to use jOOQ's parser and transform the expression tree themselves.

> EDIT: I almost forgot. One company I interviewed with a year ago said they're using jOOQ full-on, but they do not like it because of its performance

Did they get in touch? There had been a few low hanging fruit in the past, such as caching reflection lookups. I'm pretty sure that these things are edge cases. 90% of all jOOQ users don't care about this, but I'd love for jOOQ to excel also for the 10% - except, I don't know what they're doing!

So, yes, please do ask. Any feedback is greatly appreciated!

**Comment by lukaseder** [comment]
By the way, regarding the size of this project, there's a big caveat that I don't know how to solve yet. Whether the DSL should become immutable or stay partially mutable as it is today. There has been a ton of feedback to this question here:
https://twitter.com/lukaseder/status/1265708804001456129

Not everyone agrees the DSL should become immutable. Specifically, this could lead to very nasty regressions. So, to be able to cache SQL in the best way, immutability would be needed, and it would be a very tricky project with a lot of risk.

A simpler cache may be possible but it would be less elegant.

**Comment by JanecekPetr** [comment]
> So, yes, please do ask. Any feedback is greatly appreciated!

My friends inside haven't really heard any complaints either, as expected the vast majority of time is spent in the DB crunching through their monstrous queries. I guess we can safely assume that one case was a one-off, or a noisy signal.

> Whether the DSL should become immutable or stay partially mutable as it is today

I saw the thread and do not have a clear opinion, so I did not chime in. It's a trade-off either way, good luck finding the correct design :)

**Comment by lukaseder** [comment]
> I guess we can safely assume that one case was a one-off, or a noisy signal.

Yes, it's not the usual case, but as you say, it's a noisy one. Every benchmark I've seen so far was a half assed benchmark using a quick and dirty H2 setup, in case of which jOOQ doesn't shine as much as I'd like it to.

I do think that in those cases, I need to find an answer, with the side effect of faring better in those benchmarks...

> good luck finding the correct design :)

It will be discovered. Good things come to those who wait!
